File: python/canvastekk_workflow_sdk/workflow/executor.py
"""
Node Executor Strategy

Defines how a node is executed: in-process (BaseNode.execute) or HTTP (POST /execute).
"""

from __future__ import annotations

import asyncio
from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Any

import httpx

if TYPE_CHECKING:
    from canvastekk_workflow_sdk.base import BaseNode
    from canvastekk_workflow_sdk.context import ExecutionContext


class NodeExecutor(ABC):
    """Abstract strategy for executing a workflow node."""

    @abstractmethod
    async def execute(
        self,
        slug: str,
        inputs: dict[str, Any],
        context: ExecutionContext,
    ) -> dict[str, Any]:
        """Execute a node and return its outputs."""

    @abstractmethod
    def has(self, slug: str) -> bool:
        """Check if this executor can handle the given slug."""


class InProcessExecutor(NodeExecutor):
    """Execute nodes by calling BaseNode.execute() directly via asyncio.to_thread.

    Example::

        executor = InProcessExecutor()
        executor.register("segmentation-v1.0.0", MySegmentNode())
        runner = WorkflowRunner(executor)
    """

    def __init__(self) -> None:
        """Initialize in-process executor with empty node registry."""
        self._registry: dict[str, BaseNode] = {}

    def register(self, slug: str, node: BaseNode) -> InProcessExecutor:
        """Register a BaseNode instance for a slug.

        Args:
            slug: Node type slug (e.g. ``"segmentation-v1.0.0"``).
            node: BaseNode instance.

        Returns:
            self for chaining.
        """
        self._registry[slug] = node
        return self

    async def execute(
        self,
        slug: str,
        inputs: dict[str, Any],
        context: ExecutionContext,
    ) -> dict[str, Any]:
        """Execute a node by calling BaseNode.execute() in a thread.

        Args:
            slug: Node type slug.
            inputs: Input values for the node.
            context: Execution context.

        Returns:
            Output values from the node.
        """
        node = self._registry[slug]
        return await asyncio.to_thread(node.execute, inputs, context)

    def has(self, slug: str) -> bool:
        """Check if a slug is registered.

        Args:
            slug: Node type slug.

        Returns:
            True if slug is in the registry.
        """
        return slug in self._registry


class HttpExecutor(NodeExecutor):
    """Execute nodes by POSTing to their ``/execute`` endpoint.

    Posts a ``NodeExecutionRequest``-compatible payload:
    ``{"run_id": ..., "node_id": ..., "inputs": ...}``

    Example::

        executor = HttpExecutor()
        executor.register_url("segmentation-v1.0.0", "http://localhost:8001")
        runner = WorkflowRunner(executor)
    """

    def __init__(self, *, timeout: float = 300.0, retries: int = 2) -> None:
        """Initialize HTTP executor.

        Args:
            timeout: Request timeout in seconds.
            retries: Retry attempts for transient failures (connection
                errors, 502/503/504) with exponential backoff.
        """
        self._urls: dict[str, str] = {}
        self._timeout = timeout
        self._retries = retries
        # Shared client: connection pooling + keep-alive instead of a new
        # TCP+TLS handshake per node execution (DA-1711 4.3).
        self._client: httpx.AsyncClient | None = None

    def register_url(self, slug: str, url: str) -> HttpExecutor:
        """Register a base URL for a slug.

        Args:
            slug: Node type slug (e.g. ``"segmentation-v1.0.0"``).
            url: Base URL where the node is hosted (e.g. ``"http://localhost:8001"``).

        Returns:
            self for chaining.
        """
        self._urls[slug] = url.rstrip("/")
        return self

    async def execute(
        self,
        slug: str,
        inputs: dict[str, Any],
        context: ExecutionContext,
    ) -> dict[str, Any]:
        """Execute a node by POSTing to its HTTP endpoint.

        Args:
            slug: Node type slug.
            inputs: Input values for the node.
            context: Execution context.

        Returns:
            Output values from the node.

        Raises:
            RuntimeError: If the node returns failure status.
        """
        url = self._urls[slug]
        payload = {
            "run_id": context.run_id,
            "node_id": context.node_id,
            "inputs": inputs,
        }
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=self._timeout)
        client = self._client

        response: httpx.Response | None = None
        for attempt in range(self._retries + 1):
            try:
                response = await client.post(f"{url}/execute", json=payload)
                response.raise_for_status()
                break
            except (httpx.TransportError, httpx.HTTPStatusError) as exc:
                transient = isinstance(exc, httpx.TransportError) or (
                    exc.response.status_code in (502, 503, 504)
                )
                if not transient or attempt == self._retries:
                    raise
                await asyncio.sleep(0.5 * (2**attempt))

        assert response is not None
        data = response.json()
        if not isinstance(data, dict):
            raise RuntimeError(f"Node '{slug}' returned malformed response (not an object)")
        if data.get("status") == "pass":
            outputs = data.get("outputs", {})
            if not isinstance(outputs, dict):
                raise RuntimeError(f"Node '{slug}' returned malformed outputs (not an object)")
            return outputs
        raise RuntimeError(
            f"Node '{slug}' returned failure: {data.get('error', 'unknown')}"
        )

    def has(self, slug: str) -> bool:
        """Check if a slug is registered.

        Args:
            slug: Node type slug.

        Returns:
            True if slug is in the registry.
        """
        return slug in self._urls

File: typescript/src/workflow/executor.ts
import type { ExecutionContext } from "../context.js";
import { NodeExecutionResponseSchema } from "../response.js";

/**
 * Abstract base for node execution strategies.
 */
export abstract class NodeExecutor {
  /**
   * Executes a node by slug.
   * @param slug - Node slug
   * @param inputs - Node inputs
   * @param context - Execution context
   * @returns Node outputs
   */
  abstract execute(
    slug: string,
    inputs: Record<string, unknown>,
    context: ExecutionContext,
  ): Promise<Record<string, unknown>>;
  /** Checks if a node is registered for the given slug. */
  abstract has(slug: string): boolean;
}

/**
 * Executes nodes in-process without HTTP calls.
 */
export class InProcessExecutor extends NodeExecutor {
  private _registry: Map<string, { execute: (inputs: Record<string, unknown>, ctx: ExecutionContext) => Record<string, unknown> | Promise<Record<string, unknown>> }> = new Map();

  /**
   * Registers a node executor function.
   * @param slug - Node slug
   * @param node - Node with execute method
   * @returns This executor for chaining
   */
  register(slug: string, node: { execute: (inputs: Record<string, unknown>, ctx: ExecutionContext) => Record<string, unknown> | Promise<Record<string, unknown>> }): this {
    this._registry.set(slug, node);
    return this;
  }

  /**
   * Executes a node in-process by calling its execute() method directly.
   * @param slug - Node slug
   * @param inputs - Node inputs
   * @param context - Execution context
   * @returns Node outputs
   * @throws {Error} If no node is registered for the slug
   */
  override async execute(slug: string, inputs: Record<string, unknown>, context: ExecutionContext): Promise<Record<string, unknown>> {
    const node = this._registry.get(slug);
    if (!node) throw new Error(`No node registered for slug '${slug}'`);
    return node.execute(inputs, context);
  }

  /** Checks if a node is registered for the given slug. */
  has(slug: string): boolean {
    return this._registry.has(slug);
  }
}

/**
 * Executes nodes via HTTP requests to remote services.
 */
export class HttpExecutor extends NodeExecutor {
  private _urls: Map<string, string> = new Map();
  private _timeout: number;

  /**
   * Creates a new HTTP executor.
   * @param opts - Executor options
   */
  constructor(opts?: { timeout?: number }) {
    super();
    this._timeout = opts?.timeout ?? 300_000;
  }

  /**
   * Registers a node URL.
   * @param slug - Node slug
   * @param url - Node base URL
   * @returns This executor for chaining
   */
  registerUrl(slug: string, url: string): this {
    this._urls.set(slug, url.replace(/\/$/, ""));
    return this;
  }

  /**
   * Executes a node by sending an HTTP POST to its /execute endpoint.
   * @param slug - Node slug
   * @param inputs - Node inputs
   * @param context - Execution context
   * @returns Node outputs
   * @throws {Error} If no URL is registered or the node returns an error
   */
  override async execute(slug: string, inputs: Record<string, unknown>, context: ExecutionContext): Promise<Record<string, unknown>> {
    const url = this._urls.get(slug);
    if (!url) throw new Error(`No URL registered for slug '${slug}'`);

    const payload = {
      run_id: context.runId,
      node_id: context.nodeId,
      inputs,
    };

    const resp = await fetch(`${url}/execute`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(this._timeout),
    });

    if (!resp.ok) {
      throw new Error(`Node '${slug}' returned HTTP ${resp.status}`);
    }

    // Validate the remote response against the wire schema — a null/array
    // body previously crashed with a TypeError instead of a clean error.
    const parsed = NodeExecutionResponseSchema.safeParse(await resp.json());
    if (!parsed.success) {
      throw new Error(`Node '${slug}' returned malformed response: ${parsed.error.issues[0]?.message ?? "schema mismatch"}`);
    }
    if (parsed.data.status === "pass") {
      return parsed.data.outputs ?? {};
    }
    throw new Error(`Node '${slug}' returned failure: ${parsed.data.error ?? "unknown"}`);
  }

  /** Checks if a URL is registered for the given slug. */
  has(slug: string): boolean {
    return this._urls.has(slug);
  }
}
